pyop2/backends/cuda.py

```diff
@@ -0,0 +1,625 @@
+    def data(self):
+
+        if len(self._data) == 0:
+            raise RuntimeError("Illegal access: No data associated with"
+                               " this Global!")
+
+        if cuda_backend.offloading:
+            self.ensure_availability_on_device()
+            self._availability_flag = AVAILABLE_ON_DEVICE_ONLY
+            return self._cuda_data
+        else:
+            self.ensure_availability_on_host()
+            self._availability_flag = AVAILABLE_ON_HOST_ONLY
+            return self._data
+
+    @property
+    def data_ro(self):
+        if len(self._data) == 0:
+            raise RuntimeError("Illegal access: No data associated with"
+                               " this Global!")
+
+        if cuda_backend.offloading:
+            self.ensure_availability_on_device()
+            view = self._cuda_data.view()
+            # FIXME: PyCUDA doesn't support read-only arrays yet
+            return view
+        else:
+            self.ensure_availability_on_host()
+            view = self._data.view()
+            view.setflags(write=False)
+            return view
+
+
+@dataclass(frozen=True)
+class CUFunctionWithExtraArgs:
+    cu_func: cuda.Function
+    args_to_append: Tuple[cuda.DeviceAllocation, ...]
+
+    def __call__(self, grid, block, *args):
+        if all(grid + block):
+            self.cu_func.prepared_call(grid, block,
+                                       *args,
+                                       *[arg_to_append.gpudata
+                                         for arg_to_append in self.args_to_append])
+
+
+class GlobalKernel(AbstractGlobalKernel):
+    @classmethod
+    def _cache_key(cls, *args, **kwargs):
+        key = super()._cache_key(*args, **kwargs)
+        key = key + (configuration["gpu_strategy"],)
+        return key
+
+    @utils.cached_property
+    def encoded_cache_key(self):
+        return md5(str(self.cache_key[1:]).encode()).hexdigest()
+
+    @utils.cached_property
+    def computational_grid_expr_cache_file_path(self):
+        cachedir = configuration["cache_dir"]
+        return os.path.join(cachedir, f"{self.encoded_cache_key}_grid_params")
+
+    @utils.cached_property
+    def extra_args_cache_file_path(self):
+        cachedir = configuration["cache_dir"]
+        return os.path.join(cachedir, f"{self.encoded_cache_key}_extra_args.npz")
+
+    @memoize_method
+    def get_grid_size(self, start, end):
+        from pymbolic import parse, evaluate
+
+        fpath = self.computational_grid_expr_cache_file_path
+
+        with open(fpath, "r") as f:
+            glens, llens = f.read().split("\n")
+
+        glens = parse(glens)
+        llens = parse(llens)
+
+        parameters = {"start": start, "end": end}
+
+        grid = tuple(int(evaluate(glens[i], parameters)) if i < len(glens) else 1
+                     for i in range(2))
+        block = tuple(int(evaluate(llens[i], parameters)) if i < len(llens) else 1
+                      for i in range(3))
```

Comment:
> This cached data is generated by a computer

I preferred a human-readable format to keep the data valid across pyop2-versions which might not be the case with pickling. (Happy to serialize in a different format)

> no need to parse and evaluate it,

I don't think we can avoid the `evaluate`s because the grid sizes are typically symbolic expressions of the form: `(end - start) // 32 + 1`. Therefore the grid sizes are evaluated at runtime during a `GlobalKernel.__call__` invocation.

Reply:
Pickle version incompatibilities are fine (they would appear when you upgrade python and we don't guarantee very much about the portability of on disk cached data between `firedrake-update` calls, let alone python version upgrades).

What you're attempting to do is cache a function that computes these things. So in the `code_to_compile` case, just write that function, and pickle it.

Although is this a consequence of the on-disk kernel/compiled code not remembering the loopy representation?

Reply:
> Pickle version incompatibilities are fin

Cool, will change. (Thanks!)

> Although is this a consequence of the on-disk kernel/compiled code not remembering the loopy representation?

Yes, the compiled code stores only the plain CUDA/OpenCL device kernel. Loopy's cache on disk mechanism works by storing both the device kernel and the host code that would call end up calling this device kernel. But adapting PyOP2 to call loopy kernels would be a bit disruptive and I _think_ loopy's cache dir being `$XDG_CACHE_DIR` has run into some problems in the past.

Reply:
> just write that function,

Planning to simply generate a python function corresponding to the expression: https://github.com/inducer/pymbolic/pull/107/

Reply:
This discussion has echos of https://github.com/inducer/loopy/pull/650#discussion_r936065984 to me, in the sense that it's possibly indicative of a  missing "invoker" level, where these (presumably) grid sizes likely should reside.

Reply:
Does that sound like a potentially reasonable thing to pursue? How are these grid sizes used?

Reply:
> that it's possibly indicative of a missing "invoker" level

Yes, there is indeed a need for the python invoker code so that we are able to call all possible loopy kernels, for example, kernels with global barriers or temporary variables in global address space. However, the current set of transformations on a PyOP2-generated kernel are relatively simple and do not make use of such features.

> How are these grid sizes used?

During compilation, we cache the python function that is generated from inducer/pymbolic#107 on disk. And whenever we intend to launch a pycuda/pyopencl kernel we `exec` the generated python code-block and compute the grid sizes. An example of the generated python function that is cached on disk looks like this:
```python
def get_grid_sizes(*, end, start):
    return ((-1 * start + ((31 + end) // 32 + (31 + 31 * start) // 32),), (32,))
```

Reply:
For my education, how come you chose to not go with the generated invokers?

Reply:
For PyOpenCL we could borrow the invoker, but for PyCUDA/C-target we don't have an invoker in loopy (yet).

Reply:
- Also, for the pyopencl invoker, a minor concern is that the disk-caching semantics don't play well with PyOP2's disk-caching semantics.
- I feel adding an invoker for the PyCUDA target shouldn't be too difficult. Taking a look at it.

Reply:
I am working on the PyCUDA/PyOpenCL invokers. For the C-target, reorganizing `pyop2/compilation.py` is a bigger undertaking and also out of the scope for this PR.